### PR TITLE
add support to run dl_canvas_unittests on GPU surfaces (off by default)

### PR DESCRIPTION
**Pull request by flar**
The primary function of this PR is to enable running the canvas_unittests rendering compliance tests on GPU surfaces. Primarily these tests verify that we record the proper information inside a DisplayList that lets us accurately reproduce the rendering stream from which they were recorded, so it isn't strictly necessary to run these tests on all output surface types - as a result this PR doesn't actually change the default running mode which is still on Software surfaces which supports all possible testing environments (some of the dither tests must run on 565 surfaces to invoke dithering and our test-surface creation code doesn't uniformly support those surfaces (yet) on all GPU backends.

Some side effects of this PR include:
- generalizing the surface creation that was previously only used for the DL benchmarks for use in more testing areas
- moving the display list sources to a more hierarchical source tree by moving the new support code into a testing/ sub-directory and also to make their types shorter by abbreviating the common class name prefix to `Dl`

The next step will be to create a test mechanism that allows us to use Impeller to perform the rendering. At that point we can use this test harness, or something very much like it, to verify that Impeller rendering output is similar to the legacy Flutter output that has been generated by Skia.

### CODE REVIEW

**Comment by zanderso** [comment]
From PR review triage: It looks like this is WIP, so I'll update labels to match.

**Comment by flar** [comment]
> From PR review triage: It looks like this is WIP, so I'll update labels to match.

No. The review requests were sent out over a week ago and ignored.

**Comment by flar** [comment]
Note that someone recently turned on `impeller_enable_vulkan` on Mac which disabled our support for GL test surfaces there as per the comment in `display_list/testing/BUILD.gn`

File: display_list/display_list_canvas_unittests.cc [hunk 1]
```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
...
 
  private:
-  sk_sp<SkSurface> surface_;
   SkPixmap pixmap_;
   void* addr_ = nullptr;
+  SkMatrix matrix_;
+  SkIRect clip_bounds_;
+};
+
+#define RENDER_JOB_BASE                 \
```
Comment (chinmaygarde) [inline display_list/display_list_canvas_unittests.cc:283]: Why couldn't this just be a base struct struct?

File: display_list/display_list_canvas_unittests.cc [hunk 2]
```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
...
-    ref_matrix_ = ref_canvas()->getTotalMatrix();
-    ref_clip_ = ref_canvas()->getDeviceClipBounds();
-    cv_renderer(ref_canvas(), paint);
-    ref_pixmap_ = ref_surface_->pixmap();
   }
 
-  const SkImageInfo& info() const { return info_; }
-  SkCanvas* ref_canvas() { return ref_surface_->canvas(); }
+  template <typename J>
+  std::unique_ptr<RenderResult> getResult(J& job) const {
```
Comment (chinmaygarde) [inline display_list/display_list_canvas_unittests.cc:386]: I may be missing something but I am not sure why this needs to be templated if the jobs can subclass a common struct.

File: display_list/display_list_canvas_unittests.cc [hunk 3]
```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
...
   const DisplayListBuilder& ref_attr() const { return ref_attr_; }
-  const SkMatrix& ref_matrix() const { return ref_matrix_; }
-  const SkIRect& ref_clip_bounds() const { return ref_clip_; }
-  const SkPixmap* ref_pixmap() const { return ref_pixmap_; }
+  const RenderResult* ref_result() const { return ref_result_.get(); }
 
  private:
-  explicit RenderEnvironment(const SkImageInfo& info) : info_(info) {
-    ref_surface_ = MakeSurface();
+  static bool matches(const DlSurfaceInstance* surface, int w, int h) {
```
Comment (chinmaygarde) [inline display_list/display_list_canvas_unittests.cc:419]: Nit: `SizeMatches` maybe (go/fluttercppl#Function_Names)? Though I suppose mixing this style and Skia style is common here so perhaps fine for consistency.

File: display_list/display_list_canvas_unittests.cc [hunk 4]
```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
...
  public:
-  static RenderEnvironment Make565() {
-    return RenderEnvironment(
-        SkImageInfo::Make({1, 1}, kRGB_565_SkColorType, kOpaque_SkAlphaType));
+  RenderEnvironment(const DlSurfaceProvider* provider, PixelFormat format)
+      : provider_(provider), format_(format) {
+    if (provider->supports(format)) {
+      main_surface_ =
+          provider->MakeOffscreenSurface(kTestWidth, kTestHeight, format);
+      alt_surface_ = provider->MakeOffscreenSurface(kTestWidth * 2,
```
Comment (chinmaygarde) [inline display_list/display_list_canvas_unittests.cc:355]: Can't tell how or why this alt surface exists. Can you elaborate?

File: display_list/testing/dl_test_surface_software.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#define FLUTTER_DISPLAY_LIST_TESTING_DL_TEST_SURFACE_SOFTWARE_H_
+
+#include "flutter/display_list/testing/dl_test_surface_provider.h"
+
+namespace flutter {
+namespace testing {
+
+class DlSoftwareSurfaceProvider : public DlSurfaceProvider {
+ public:
+  DlSoftwareSurfaceProvider() : DlSurfaceProvider() {}
```
Comment (chinmaygarde) [inline display_list/testing/dl_test_surface_software.h:15]: Nit: You can just default this.

File: display_list/testing/dl_test_surface_software.cc [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+namespace flutter {
+namespace testing {
+
+using PixelFormat = DlSurfaceProvider::PixelFormat;
+
+bool DlSoftwareSurfaceProvider::InitializeSurface(size_t width,
+                                                  size_t height,
+                                                  PixelFormat format) {
+  primary_ = MakeOffscreenSurface(width, height, format);
+  return true;
```
Comment (chinmaygarde) [inline display_list/testing/dl_test_surface_software.cc:19]: Maybe `!!primary_`? When would initialization ever fail?

**Review by chinmaygarde — COMMENTED** [review]
I guess the bit I didn't understand is the templating scheme and use of the DIY inheritance using macros. Otherwise, mostly just nits and questions.

File: display_list/display_list_canvas_unittests.cc (hunk 1 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:283]: I wanted to use struct initializers throughout at the points of invocation, but that style of initialization cannot work with a v-table. So, no virtual methods -> no struct initializers -> complicated multi-argument constructors with arguments that are only used in some cases -> harder to read invocation code.

File: display_list/display_list_canvas_unittests.cc (hunk 2 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:386]: Struct initializers prevent virtual methods.

File: display_list/display_list_canvas_unittests.cc (hunk 3 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:419]: At one point they were constructed as needed and could be null so the test was more than just the size, but then I went with just hand-coding 2 different sizes as I know that's all the tests use and so this function isn't really useful any more.

File: display_list/display_list_canvas_unittests.cc (hunk 4 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:355]: I run tests in SIZE and SIZE*2 to test dispatch-time-scaling, so the main surface is created at the normal size and the alt is created at double size.

At one point I was going to generalize this and allow the surfaces to be created dynamically depending on the request size, but it got complicated and I realized that it was far simpler to just create the 1x and 2x surfaces in the constructor.

The fields could stand to be renamed and the getSurface() method simplified after that decision.

File: display_list/testing/dl_test_surface_software.cc (hunk 6 above)
Comment (flar) [reply display_list/testing/dl_test_surface_software.cc:19]: OOM? It doesn't have much configuration to get wrong with the SW provider.

**Review by chinmaygarde — APPROVED** [review]

**Comment by flar** [comment]
BTW, I'm reworking the Job structs so that there is an attribute struct which can have named initializers and a Renderer struct which has virtual methods so that code can have the best of both worlds.

File: display_list/display_list_canvas_unittests.cc (hunk 1 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:283]: I turned this into 2 structs - non-virtual `RenderJobInfo` with the optional attributes and named struct initialization, and a virtual `JobRenderer` which hosts the necessary Render() method, along with some specific methods used to retrieve task-specific info during the testing. No more templates or macros.

File: display_list/display_list_canvas_unittests.cc (hunk 2 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:386]: Fixed as mentioned above.

File: display_list/display_list_canvas_unittests.cc (hunk 3 above)
Comment (flar) [reply display_list/display_list_canvas_unittests.cc:419]: Changed names to `surface_1x_` and `surface_2x_` to make them more obvious.

File: display_list/testing/dl_test_surface_software.h (hunk 5 above)
Comment (flar) [reply display_list/testing/dl_test_surface_software.h:15]: Done

File: display_list/testing/dl_test_surface_software.cc (hunk 6 above)
Comment (flar) [reply display_list/testing/dl_test_surface_software.cc:19]: Done, but the others don't really test for success in creating the surface beyond validating the arguments and pixel format against their implementation capabilities...?

**Comment by flar** [comment]
@chinmaygarde if you are good with my [last commit](https://github.com/flutter/engine/pull/38993/commits/a741f147a8b710283e43febd3c17fc058962246b), feel free to add the autosubmit label...

**Pull request merged**